Frameworks/Foundation/NSOperation.mm

```diff
@@ -1,115 +1,112 @@
-
-*/
+//******************************************************************************
+//
+// Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
 
 #include "Starboard.h"
 #include "StubReturn.h"
 #include "Foundation/NSOperation.h"
-#include "Foundation/NSString.h"
-#include "Foundation/NSMutableArray.h"
-#include "LoggingNative.h"
 
 static const wchar_t* TAG = L"NSOperation";
 
-#if __cplusplus
-#include <pthread.h>
-#include <string.h>
-struct NSOperationPriv {
-    NSOperationQueuePriority priority;
-    id dependencies;
-    void (^completionBlock)(void);
+@implementation NSOperation {
+    StrongId<NSMutableArray> _dependencies;
+    void (^_completionBlock)(void);
+    pthread_cond_t _finishCondition;
+    pthread_mutex_t _finishLock;
+    pthread_mutex_t _dependenciesLock;
+    pthread_mutex_t _completionBlockLock;
+}
 
-    pthread_cond_t finishCondition;
-    pthread_mutex_t finishLock;
+@synthesize cancelled = _cancelled;
+@synthesize executing = _executing;
+@synthesize finished = _finished;
 
-    int executing : 1;
-    int cancelled : 1;
-    int finished : 1;
+/**
+ @Status Interoperable
+*/
++ (BOOL)automaticallyNotifiesObserversForKey:(NSString*)key {
+    // This class dispatches its own notifications.
+    return NO;
+}
 
-    NSOperationPriv() {
-        memset(this, 0, sizeof(NSOperationPriv));
-        pthread_cond_init(&finishCondition, 0);
+- (void)observeValueForKeyPath:(NSString*)keyPath ofObject:(id)object change:(NSDictionary*)change context:(void*)context {
+    if ([keyPath isEqualToString:@"isFinished"]) {
+        pthread_mutex_lock(&_finishLock);
+        pthread_cond_broadcast(&_finishCondition);
+        pthread_mutex_unlock(&_finishLock);
+    }
+}
 
+- (id)init {
+    if (self = [super init]) {
+        _dependencies.attach([NSMutableArray new]);
+        pthread_cond_init(&_finishCondition, 0);
         pthread_mutexattr_t attr;
-
         pthread_mutexattr_init(&attr);
         pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
-        pthread_mutex_init(&finishLock, &attr);
+        pthread_mutex_init(&_finishLock, &attr);
         pthread_mutexattr_destroy(&attr);
-    }
+        pthread_mutex_init(&_dependenciesLock, NULL);
+        pthread_mutex_init(&_completionBlockLock, NULL);
```

Comment:
we talked earlier about using NSLock and yep it's not going to be efficient as just using raw pthreads since it objc_msgSend will add addition overhead.

Although pthreads are fine, we have allot going on here, it's easier to forget to set/destroy something. Since we support C++ on these files, we should go ahead and use std::mutex, recursive mutex, etc. 

This would make the code much cleaner to maintain and read, also avoid pitfalls that we might encounter in the future maintenance.  Going forward we should use them.